### PR TITLE
Resurrect Ionwf

### PR DESCRIPTION
**Pull request by Paul-St-Young**
Ratio test is broken at the current commit a7047891a7191ffa3b772905d5043862748d167f
However ratio test passes as of Dec. 23 commit 8194d94148ec6a2bb60e44c349b9b1e6ca152c09
 
update: Ratio test is still broken at commit 8194d94148ec6a2bb60e44c349b9b1e6ca152c09

### CODE REVIEW

**Comment by prckent** [comment]
OK to test

File: src/QMCWaveFunctions/IonOrbitalBuilder.cpp [hunk 1]
```diff
@@ -28,6 +28,8 @@ bool
 IonOrbitalBuilder:: put(xmlNodePtr cur)
 {
   ParticleSet &p = targetPtcl;
+  widthOpt.resize(targetPtcl.getTotalNum());
+  widthOpt = -1;
```
Comment (ye-luo) [inline src/QMCWaveFunctions/IonOrbitalBuilder.cpp]: Why initialize the value to -1? The later non-zero checking seems assume initialized value 0 and requires full user input.

**Comment by ye-luo** [comment]
@Paul-St-Young I would like to understand why my change on develop break the IonOrbital builder. Since you provide a fix, could you explain your logic?

File: manual/ionwf.tex [hunk 2]
```diff
@@ -0,0 +1,71 @@
+\section{Gaussian Product Wavefunction}
+\label{sec:ionwf}
+
+The Gaussian Product wavefunction implements eq.~\ref{eq:gauss_prod_wf}
```
Comment (prckent) [inline manual/ionwf.tex]: Thanks for adding the manual section. Please can you add a citation to any published work using this wavefunction (QMCPACK or not) and a couple of sentences on when/where the wavefunction might be used.

File: src/QMCWaveFunctions/IonOrbitalBuilder.cpp (hunk 1 above)
Comment (Paul-St-Young) [reply src/QMCWaveFunctions/IonOrbitalBuilder.cpp]: @ye-luo the width parameter is required. The intended behavior is abort when no user input is given.

**Comment by Paul-St-Young** [comment]
@ye-luo I actually have no idea what broke IonOrbital and provided no fix. I noticed that my test is broken on the current develop HEAD. I was wondering if it was my test or the code, so I went back to a previous commit.

File: manual/ionwf.tex (hunk 2 above)
Comment (Paul-St-Young) [reply manual/ionwf.tex]: @prckent unfortunately I have no knowledge of the history of this wavefunction. The file banner suggests it was first developed by Ken Esler. Could we contact him?

Comment (Paul-St-Young) [reply manual/ionwf.tex]: I will add a few sentences describing potential use of the wavefunction.

File: src/QMCWaveFunctions/IonOrbitalBuilder.cpp (hunk 1 above)
Comment (ye-luo) [reply src/QMCWaveFunctions/IonOrbitalBuilder.cpp]: I agree that the intended behaviour is abort when no user input is given. But I'm not convinced that it justifies your change. The old code relies on oAttrib to do the resize of widthOpt if a user input is give. If no input is given. widthOpt has zero size and abort. In your change, you resize and initialize the value to -1. The following num_nonzero will not be able to abort if no user input is given. If you set the initial value to 0, I may agree.

Comment (Paul-St-Young) [reply src/QMCWaveFunctions/IonOrbitalBuilder.cpp]: @ye-luo "The old code relies on oAttrib to do the resize of widthOpt" is exactly why IonOrbital has been broken. For some reason `oAttrib` no longer automatically does the resize. I see your point about the change in initial default now. I will change the defaults to 0.

**Comment by ye-luo** [comment]
After my investigation, this is not a recent breakage. v3.1.1 also has the size issue. @Paul-St-Young your magic resize is the correct fix. I will manually do the merge to make your branch up-to-date after PR #622 and make the two lines in 1.

**Comment by ye-luo** [comment]
@Paul-St-Young 
```
Finite difference test: PASS
Wavefunction ratio exceeds tolerance 1.0000000000e-03) for particle 0
  Internally computed ratio = 2.1626500948e-01
  Separately computed ratio = 2.1626476619e-01
Ratio test: FAIL
```
Is this failure normal or just a false alarm?

**Comment by Paul-St-Young** [comment]
@ye-luo thanks for looking into this. I am not sure if what you are seeing is a false alarm or not. Do you see failure using the `check_grad_lap.py` script?

Please also move the quantum particle around and rerun the wavefunction test to see if the error persists.

**Comment by ye-luo** [comment]
@Paul-St-Young I ran the ctest and got a failure. Then I opened the log and saw the error.

**Comment by Paul-St-Young** [comment]
@ye-luo the ctest should pass. Could you try an older binary?

**Comment by ye-luo** [comment]
@Paul-St-Young what do you mean by older binary?

**Comment by Paul-St-Young** [comment]
Sorry, I meant an older commit, like 8194d94148ec6a2bb60e44c349b9b1e6ca152c09

**Comment by Paul-St-Young** [comment]
No need to port the entire ctest over. I just slap on the MNC, run `grad_lap.xml`, then `check_grad_lap.py`.

**Comment by ye-luo** [comment]
I tried the old commit and got
```
Wavefunction ratio exceeds tolerance 1.0000000000e-03) for particle 0
  Internally computed ratio = 8.5525616287e-02
  Separately computed ratio = 8.5525038991e-02
Ratio test: FAIL
```
I just look at the number and it seems that the relative error is only 0.00067, smaller than 0.001.

**Comment by Paul-St-Young** [comment]
@ye-luo Does the error persist when you move the quantum particle around?

**Comment by Paul-St-Young** [comment]
Oh I am sorry, the old commit is also broken. I must have used my own binary when I did my test.
Please try move the quantum particle to `0.1 -0.1 0.2`. It will make the error more obvious.

**Comment by ye-luo** [comment]
I move the particle in particleset name="e"
```
Finite difference test: PASS
Wavefunction ratio exceeds tolerance 1.0000000000e-03) for particle 0
  Internally computed ratio = 4.1884301184e+01
  Separately computed ratio = 8.4155461651e-02
Ratio test: FAIL
```
So?

**Comment by Paul-St-Young** [comment]
The internal ratio 48.9 differs from finite-difference ratio 0.084, so the wavefunction ratio is wrong.
This is the ratio test failure I mentioned at the beginning of this PR. Do you have an idea what caused it?

**Comment by Paul-St-Young** [comment]
The code is still buggy. The ratio test passes on my branch.

**Comment by ye-luo** [comment]
Which is 'my branch'?
A quick look to the code indicates it is not correct in many places. The orbital should be like the one body Jastrow. The distance table handling seems not right. The Orbital class should only hold the index and then get the table from the argument P.

**Comment by Paul-St-Young** [comment]
https://github.com/Paul-St-Young/qmcpack/tree/kylin

**Comment by Paul-St-Young** [comment]
Oh shoot, I forgot a piece of fix when I made this PR. The ratio test should be fixed now. Sorry about the confusion @ye-luo

File: src/QMCWaveFunctions/IonOrbital.cpp [hunk 3]
```diff
@@ -155,7 +156,12 @@ IonOrbital::ratioGrad(ParticleSet& P, int iat, GradType& grad_iat)
 }
 
 void
-IonOrbital::restore(int iat) {}
+IonOrbital::restore(int iat)
```
Comment (ye-luo) [inline src/QMCWaveFunctions/IonOrbital.cpp]: Restore should not be left empty. evaluateLog should directly assign zero.

Comment (Paul-St-Young) [reply src/QMCWaveFunctions/IonOrbital.cpp]: Yes, that was the problem. Please see "fix ratio test". I accidentally left out this piece of the fix when I made this PR. Sorry about that.

Comment (ye-luo) [reply src/QMCWaveFunctions/IonOrbital.cpp]: I saw "fix ratio test". I mean revert the change of restore and assign zero directly in evaluateLog. One goal is to remove all the restore at least in CPU code.

Comment (Paul-St-Young) [reply src/QMCWaveFunctions/IonOrbital.cpp]: Oh I see. I was not aware of that goal. Will clean now

**Comment by prckent** [comment]
Many thanks @Paul-St-Young

**Pull request merged**